input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing Undo and Redo key combinations with an existing history');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        // Decide whether to use  Key.COMMAND (mac) or Key.CONTROL (everything else)
+        const modifierKey = navigator.userAgent.toLowerCase().includes('mac') ? '\u2318' : '\uE009';
+
+        // Undo
+        await new test_driver.Actions()
+            .keyDown(modifierKey)
+            .keyDown('z')
+            .keyUp('z')
+            .keyUp(modifierKey)
+            .send();
+        // Redo
+        await new test_driver.Actions()
+            .keyDown(modifierKey)
+            .keyDown('\uE008') // Shift
+            .keyDown('z')
+            .keyUp('z')
+            .keyUp('\uE008')
+            .keyUp(modifierKey)
+            .send();
+
+        assert_equals(inputEventsLog.length, 4);
+        const inputTypes = ['historyUndo', 'historyRedo'];
+        for (let i = 0; i < inputEventsLog.length; i += 2) {
+            const beforeInputEvent = inputEventsLog[i];
+            const inputEvent = inputEventsLog[i + 1];
+            assert_equals(beforeInputEvent.type, 'beforeinput');
+            assert_equals(inputEvent.type, 'input');
+            assert_equals(beforeInputEvent.inputType, inputTypes[i / 2]);
+            assert_equals(inputEvent.inputType, inputTypes[i / 2]);
+            assert_equals(beforeInputEvent.data, inputEvent.data);
+        }
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing Undo and Redo key combinations without an existing history');
+
+    promise_test(async () => {
+        resetEditor();
+        const expectedResult = [
+            // Pressing 'a'.
+            'plain-keydown-a',
+            'plain-keypress-a',
+            'plain-beforeinput-a-null',
+            'plain-input-a-null',
+            'plain-keyup-a',
+            // Pressing Shift-'b'.
+            'plain-keydown-B',
+            'plain-keypress-B',
+            'plain-beforeinput-B-null',
+            'plain-input-B-null',
+            'plain-keyup-B',
+            // Pressing 'c'.
+            'rich-keydown-c',
+            'rich-keypress-c',
+            'rich-beforeinput-c-null',
+            'rich-input-c-null',
+            'rich-keyup-c',
+            // Pressing Shift-'d'.
+            'rich-keydown-D',
+            'rich-keypress-D',
+            'rich-beforeinput-D-null',
+            'rich-input-D-null',
+            'rich-keyup-D',
+        ];
+        const result = [];
+
+        for (const targetId of ['plain', 'rich']) {
+            for (const eventType of ['beforeinput', 'input', 'keydown', 'keypress', 'keyup']) {
+                document.getElementById(targetId).addEventListener(eventType, event => {
+                    if (event.key === 'Shift')
+                        return;
```

Comment:
Easier on the eyes, just one line it: 

```suggestion
                    if (event.key === 'Shift') return;
```

Reply:
done